### ISSUE TITLE
Dart WebSocket Server does not pass all AutoBahn tests

### ISSUE THREAD

**Issue by azenla** [dart-lang/sdk#25317]
The common standard to test WebSocket implementations is the AutoBahn test suite.
http://autobahn.ws/testsuite/

It looks like Dart doesn't pass a decent amount of them, especially WebSocket compression related tests.

Here are the results as of commit 8a445a3f01c1d1e8f3cfb4175150b549f944bae3 on a x64 Mac release build.

http://files.directcode.io/dartwstest/

CC @butlermatt

**Comment by butlermatt** [comment]
@kaendfinger Do you have details on how you ran the Autobahn test? I'm starting to look at this today. A sample of the websock server you used?

**Comment by kaendfinger** [comment]
Bind an echoing WebSocket server to port 9001 locally.

Make sure you have Python installed.

Run:
```
pip install autobahntestsuite
wstest -m fuzzingclient
```

It should run through all the tests and generate reports.

**Comment by butlermatt** [comment]
Patch submitted for review here:
https://codereview.chromium.org/1579343003/

**Comment by sgjesse** [comment]
Great with the compression part fixed.

@kaendfinger <https://github.com/kaendfinger> the running instructions you
posted only tests the server side as far as I can see. Is there a way to
test the client as well - I assume that it would require writing the client
part of the test suite in Dart.

---
Søren Gjesse
Software Engineer, Google Denmark
CVR nr. 28 86 69 84

**Comment by butlermatt** [comment]
It took me a little to get it figured out but I'm currently running the client side tests for this now

**Comment by butlermatt** [comment]
Just to follow up. The client tests were showing a couple of errors in Websocket compression. I'm running final tests on a patch now and will submit it shortly. I'll also provide repository to see the tests I used and the results from them.

**Comment by butlermatt** [comment]
Patch to resolve client issues is:
https://codereview.chromium.org/1583933007/

Test repository:
https://github.com/butlermatt/dart-autobahn-testing

Test results:
Websocket Client [Test Results](http://butlermatt.github.io/dart-autobahn-testing/clients/)
Websocket Server [Test Results](http://butlermatt.github.io/dart-autobahn-testing/servers/)

**Comment by sgjesse** [comment]
Thanks Matthew,

The patch has landed. Could you share how to run the client tests?

/Søren

**Comment by butlermatt** [comment]
Hi Søren,

As I mention in the test repository I linked to above, to run the client tests:

1) Start the wstest server: wstest -m fuzzingserver
2) Run the dart client: dart bin\wsclient.dart

Report will be generated in reports/clients/index.html from where the wstest command was run.

The wsclient.dart will first open a websocket to query the Autobahn test server that is started. Once it has the total number of tests, it will iterate through each performing a simple echo response to anything since by the server. Finally it will query to the server to generate the reports based on the test results it creates.

See the client directly here:
https://github.com/butlermatt/dart-autobahn-testing/blob/master/bin/wsclient.dart

This way when isolating on a specific set of tests, or single test with the server, the client does not need to be updated for each test.

**Comment by butlermatt** [comment]
Per comments in CL: https://codereview.chromium.org/1583933007/ I have created another patch 
https://codereview.chromium.org/1584653008/

This patch addresses 3 areas:
* Optimize client and server side no take over handling to discard encoder and decoders appropriate.
* Resolves an issue where client would not request headers based on the provided CompressionOptions at creation (and adds tests)
* Resolves an issue where server was enabling 'no context takeover' yet did not advertise it in a response and was not configured in CompressionOptions. (And adds tests)

Verified all green for websocket compression in Autobahn in 4 tests:
* Server without "no context takeover" options enabled.
* Server with "no context takeover" options enabled.
* Client without "no context takeover" options enabled.
* Client with "no context takeover" options enabled.
